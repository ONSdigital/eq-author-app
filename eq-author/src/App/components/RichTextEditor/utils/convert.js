import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = htmlToEntity => {
  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
