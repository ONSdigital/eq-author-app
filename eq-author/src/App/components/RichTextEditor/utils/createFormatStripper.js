import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig } from "App/components/RichTextEditor/entities/PipedValue";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" }
};

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [filterConfig],
    maxNesting: 0,
    whitespacedCharacters: []
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return editorState => filterEditorState(filterConfiguration, editorState);
}
