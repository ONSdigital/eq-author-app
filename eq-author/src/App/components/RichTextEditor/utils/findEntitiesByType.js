export default entityType => {
  return function findEntitiesByType(contentBlock, callback, contentState) {
    contentBlock.findEntityRanges(character => {
      const entityKey = character.getEntity();
      return (
        character.getEntity() !== null &&
        contentState.getEntity(entityKey).getType() === entityType
      );
    }, callback);
  };
};
