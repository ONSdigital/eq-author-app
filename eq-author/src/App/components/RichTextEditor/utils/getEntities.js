function getEntities(contentState, entityType) {
  const entities = [];

  contentState.getBlocksAsArray().forEach(block => {
    let selectedEntity = null;

    block.findEntityRanges(
      character => {
        const entityKey = character.getEntity();

        if (entityKey === null) {
          return false;
        }

        const entity = contentState.getEntity(entityKey);

        if (entityType && entity.getType() !== entityType) {
          return false;
        }

        selectedEntity = {
          entityKey,
          entity,
          blockKey: block.getKey()
        };

        return true;
      },
      (start, end) => {
        entities.push({ ...selectedEntity, start, end });
      }
    );
  });

  return entities;
}

export default getEntities;
