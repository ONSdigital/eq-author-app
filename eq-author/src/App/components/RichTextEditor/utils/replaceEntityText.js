import { SelectionState, Modifier } from "draft-js";

const getEntityRange = (block, entityKey) => {
  let range;

  // get the range for the entity key
  block.findEntityRanges(
    char => char.getEntity() === entityKey,
    (start, end) => {
      range = { start, end };
    }
  );

  return SelectionState.createEmpty(block.key).merge({
    anchorKey: block.key,
    anchorOffset: range.start,
    focusKey: block.key,
    focusOffset: range.end
  });
};

const replaceText = (contentState, entityKey, blockKey, text, inlineStyles) => {
  const block = contentState.getBlockForKey(blockKey);
  const entitySelection = getEntityRange(block, entityKey);

  // first replace text
  const updatedContentState = Modifier.replaceText(
    contentState,
    entitySelection,
    text,
    inlineStyles,
    entityKey
  );

  // then create new selection, which covers the updated entity
  const updatedEntitySelection = entitySelection.merge({
    focusOffset: entitySelection.getStartOffset() + text.length
  });

  // apply the updated entity to the content
  return Modifier.applyEntity(
    updatedContentState,
    updatedEntitySelection,
    entityKey
  );
};

export default replaceText;
