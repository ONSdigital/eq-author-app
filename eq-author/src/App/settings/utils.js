export const getThemeSettingsErrorCount = (questionnaire) =>
  questionnaire?.themeSettings?.validationErrorInfo?.totalCount +
    questionnaire?.validationErrorInfo?.totalCount || 0;
