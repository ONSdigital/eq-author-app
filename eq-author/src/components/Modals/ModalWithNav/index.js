import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { connect } from "react-redux";

import { gotoTab } from "redux/tabs/actions";

import Modal from "components/Modals/Modal";
import { Grid } from "components/Grid";

import Tabs from "components/Modals/ModalWithNav/Tabs";

const Dialog = styled(Modal)`
  .Modal {
    width: 60em;
    padding: 0;
  }
`;

export const UnconnectedModalWithNav = ({
  title,
  onClose,
  navItems,
  activeTabId,
  gotoTab,
  isOpen,
  id,
  ...otherProps
}) => (
  <Dialog onClose={onClose} isOpen={isOpen} {...otherProps}>
    <Grid fillHeight>
      <Tabs
        activeTabId={activeTabId}
        onChange={tabId => gotoTab(id, tabId)}
        onClose={onClose}
        navItems={navItems}
        title={title}
      />
    </Grid>
  </Dialog>
);

UnconnectedModalWithNav.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
  isOpen: PropTypes.bool,
  onClose: PropTypes.func.isRequired,
  gotoTab: PropTypes.func.isRequired,
  activeTabId: PropTypes.string,
  navItems: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      render: PropTypes.func.isRequired
    })
  )
};

UnconnectedModalWithNav.defaultProps = {
  isOpen: false
};

const mapStateToProps = (state, ownProps) => ({
  activeTabId: state.tabs[ownProps.id]
});

export default connect(
  mapStateToProps,
  { gotoTab }
)(UnconnectedModalWithNav);
