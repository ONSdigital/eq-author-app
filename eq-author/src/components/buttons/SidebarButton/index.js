import styled, { css } from "styled-components";
import { colors, radius } from "constants/theme";
import chevron from "./icon-chevron.svg";

const SidebarButton = styled.button.attrs({ role: "button" })`
  display: block;
  width: 100%;
  padding: 0.5em;
  margin-bottom: 0.5em;
  color: ${colors.text};
  border-radius: ${radius};
  border: 1px solid ${colors.bordersLight};
  text-align: left;
  font-size: 1em;
  transition: all 100ms ease-out;
  position: relative;
  cursor: pointer;
  background: ${colors.white};

  &:hover {
    border: 1px solid ${colors.borders};
    background: ${colors.lighterGrey};
  }

  &:focus {
    box-shadow: 0 0 0 3px ${colors.tertiary};
    outline: none;
  }

  &::after {
    content: "";
    display: block;
    width: 1em;
    height: 1em;
    background: url(${chevron}) no-repeat center;
    position: absolute;
    right: 0.5em;
    top: 0;
    bottom: 0;
    margin: auto;
  }

  &:disabled {
    background: ${colors.lighterGrey};
  }

  ${props =>
    props.hasError &&
    css`
      border-color: ${colors.red};
      &:focus,
      &:focus-within {
        border-color: ${colors.red};
        outline-color: ${colors.red};
        box-shadow: 0 0 0 2px ${colors.red};
      }
      &:hover {
        border-color: ${colors.red};
        outline-color: ${colors.red};
      }
    `}
`;

export const Title = styled.span`
  display: block;
  color: ${colors.darkGrey};
  font-size: 0.9em;

  &:not(:only-child) {
    margin-bottom: 0.25rem;
  }
`;

export const Detail = styled.span`
  display: block;
  color: ${colors.black};
`;

export default SidebarButton;
