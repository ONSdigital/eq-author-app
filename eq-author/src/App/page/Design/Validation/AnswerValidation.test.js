import React from "react";
import { shallow } from "enzyme";
import { render as rtlRender } from "tests/utils/rtl";

import { NUMBER, CURRENCY, PERCENTAGE, UNIT } from "constants/answer-types";
import { CENTIMETRES } from "constants/unit-types";

import SidebarButton, { Detail } from "components/buttons/SidebarButton";
import ModalWithNav from "components/modals/ModalWithNav";
import {
  UnwrappedAnswerValidation,
  validationTypes,
  MIN_INCLUSIVE_TEXT,
  MAX_INCLUSIVE_TEXT,
} from "./AnswerValidation";

const render = (props, render = shallow) => {
  return render(<UnwrappedAnswerValidation {...props} />);
};

describe("AnswerValidation", () => {
  let props;

  beforeEach(() => {
    props = {
      answer: {
        id: "1",
        type: NUMBER,
        validation: {
          minValue: {
            enabled: false,
            validationErrorInfo: { errors: [] },
          },
          maxValue: {
            enabled: false,
            validationErrorInfo: { errors: [] },
          },
        },
      },
    };
  });

  it("should render", () => {
    expect(render(props)).toMatchSnapshot();
  });

  it("should not render when there are no valid validation types", () => {
    props.answer.type = "Radio";
    expect(render(props)).toMatchSnapshot();
  });

  it("should correctly initialise state", () => {
    const wrapper = render(props);
    expect(wrapper.state("modalIsOpen")).toEqual(false);
  });

  it("should correctly update state when opening a Modals", () => {
    const wrapper = render(props);
    wrapper
      .find(SidebarButton)
      .first()
      .simulate("click");
    expect(wrapper.state("modalIsOpen")).toEqual(true);
  });

  it("should correctly update state when closing a Modals", () => {
    const wrapper = render(props);
    wrapper.find(ModalWithNav).simulate("close");
    expect(wrapper.state("modalIsOpen")).toEqual(false);
  });

  describe("validation object array", () => {
    validationTypes.forEach(validationType => {
      it(`should render the ${validationType.title} validation`, () => {
        const wrapper = shallow(validationType.render());

        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe("Numeric answer validation preview", () => {
    const NUMBER_TYPES = [PERCENTAGE, NUMBER, CURRENCY, UNIT];
    const VALIDATIONS = ["maxValue", "minValue"];
    VALIDATIONS.forEach(validation => {
      describe(validation, () => {
        it("should render custom values", () => {
          const wrapper = type => {
            const properties = {};
            if (type === UNIT) {
              properties.unit = CENTIMETRES;
            }
            return render({
              ...props,
              answer: {
                id: "1",
                type: type,
                properties,
                validation: {
                  [validation]: {
                    enabled: true,
                    custom: 5,
                    entityType: "Custom",
                  },
                },
              },
            });
          };

          NUMBER_TYPES.forEach(type => {
            expect(
              wrapper(type)
                .find(SidebarButton)
                .find(Detail)
            ).toMatchSnapshot();
          });
        });

        it("should not render when the custom value is null", () => {
          const wrapper = type =>
            render({
              ...props,
              answer: {
                id: "1",
                type: type,
                validation: {
                  [validation]: {
                    enabled: true,
                    custom: null,
                    entityType: "Custom",
                  },
                },
              },
            });

          NUMBER_TYPES.forEach(type => {
            expect(
              wrapper(type)
                .find(SidebarButton)
                .find(Detail)
            ).toMatchSnapshot();
          });
        });

        it("should render previous answer", () => {
          const wrapper = type =>
            render({
              ...props,
              answer: {
                id: "1",
                type: type,
                validation: {
                  [validation]: {
                    enabled: true,
                    previousAnswer: {
                      displayName: "foobar",
                    },
                  },
                },
              },
            });

          NUMBER_TYPES.forEach(type => {
            expect(
              wrapper(type)
                .find(SidebarButton)
                .find(Detail)
            ).toMatchSnapshot();
          });
        });

        it("should render with additional inclusive text when appropriate on all number types", () => {
          props = {
            ...props,
            answer: {
              id: "1",
              validation: {
                [validation]: {
                  enabled: true,
                  inclusive: false,
                  previousAnswer: {
                    displayName: "foobar",
                  },
                },
              },
            },
          };

          NUMBER_TYPES.forEach(type => {
            props.answer.type = type;

            const { getAllByText } = rtlRender(
              <UnwrappedAnswerValidation {...props} />
            );

            if (validation === VALIDATIONS[0]) {
              expect(
                getAllByText(`Max value ${MAX_INCLUSIVE_TEXT}`)
              ).toBeTruthy();
            }

            if (validation === VALIDATIONS[1]) {
              expect(
                getAllByText(`Min value ${MIN_INCLUSIVE_TEXT}`)
              ).toBeTruthy();
            }
          });
        });
      });
    });

    it("should render an error message when min val > max Val", () => {
      const error = [
        {
          errorCode: "ERR_MIN_LARGER_THAN_MAX",
          field: "custom",
          id: "maxValue-0183a9c0-b79f-4766-ba7a-3c3718bb9f26-custom",
          type: "validation",
          __typename: "ValidationError",
        },
      ];
      props.answer.validation.minValue.validationErrorInfo.errors = error;
      props.answer.validation.maxValue.validationErrorInfo.errors = error;

      const { getByText } = rtlRender(<UnwrappedAnswerValidation {...props} />);

      expect(
        getByText("Enter a max value that is greater than min value")
      ).toBeTruthy();
    });
  });
});
