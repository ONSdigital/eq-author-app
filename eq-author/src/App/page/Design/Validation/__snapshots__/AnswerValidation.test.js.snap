// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerValidation Numeric answer validation preview maxValue should not render when the custom value is null 1`] = `null`;

exports[`AnswerValidation Numeric answer validation preview maxValue should not render when the custom value is null 2`] = `null`;

exports[`AnswerValidation Numeric answer validation preview maxValue should not render when the custom value is null 3`] = `null`;

exports[`AnswerValidation Numeric answer validation preview maxValue should not render when the custom value is null 4`] = `null`;

exports[`AnswerValidation Numeric answer validation preview maxValue should render custom values 1`] = `
<SidebarButton__Detail>
  5%
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview maxValue should render custom values 2`] = `
<SidebarButton__Detail>
  5
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview maxValue should render custom values 3`] = `
<SidebarButton__Detail>
  £5
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview maxValue should render custom values 4`] = `
<SidebarButton__Detail>
  5 cm
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview maxValue should render previous answer 1`] = `
<SidebarButton__Detail>
  foobar
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview maxValue should render previous answer 2`] = `
<SidebarButton__Detail>
  foobar
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview maxValue should render previous answer 3`] = `
<SidebarButton__Detail>
  foobar
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview maxValue should render previous answer 4`] = `
<SidebarButton__Detail>
  foobar
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview minValue should not render when the custom value is null 1`] = `null`;

exports[`AnswerValidation Numeric answer validation preview minValue should not render when the custom value is null 2`] = `null`;

exports[`AnswerValidation Numeric answer validation preview minValue should not render when the custom value is null 3`] = `null`;

exports[`AnswerValidation Numeric answer validation preview minValue should not render when the custom value is null 4`] = `null`;

exports[`AnswerValidation Numeric answer validation preview minValue should render custom values 1`] = `
<SidebarButton__Detail>
  5%
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview minValue should render custom values 2`] = `
<SidebarButton__Detail>
  5
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview minValue should render custom values 3`] = `
<SidebarButton__Detail>
  £5
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview minValue should render custom values 4`] = `
<SidebarButton__Detail>
  5 cm
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview minValue should render previous answer 1`] = `
<SidebarButton__Detail>
  foobar
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview minValue should render previous answer 2`] = `
<SidebarButton__Detail>
  foobar
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview minValue should render previous answer 3`] = `
<SidebarButton__Detail>
  foobar
</SidebarButton__Detail>
`;

exports[`AnswerValidation Numeric answer validation preview minValue should render previous answer 4`] = `
<SidebarButton__Detail>
  foobar
</SidebarButton__Detail>
`;

exports[`AnswerValidation should not render when there are no valid validation types 1`] = `""`;

exports[`AnswerValidation should render 1`] = `
<ContextProvider
  value={
    Object {
      "answer": Object {
        "id": "1",
        "type": "Number",
        "validation": Object {
          "maxValue": Object {
            "enabled": false,
            "validationErrorInfo": Object {
              "errors": Array [],
            },
          },
          "minValue": Object {
            "enabled": false,
            "validationErrorInfo": Object {
              "errors": Array [],
            },
          },
        },
      },
    }
  }
>
  <SidebarButton
    data-test="sidebar-button-min-value"
    hasError={false}
    key="minValue"
    onClick={[Function]}
  >
    <SidebarButton__Title>
      Min value
       
    </SidebarButton__Title>
  </SidebarButton>
  <SidebarButton
    data-test="sidebar-button-max-value"
    hasError={false}
    key="maxValue"
    onClick={[Function]}
  >
    <SidebarButton__Title>
      Max value
       
    </SidebarButton__Title>
  </SidebarButton>
  <ModalWithNav
    id="modal-validation-1"
    isOpen={false}
    navItems={
      Array [
        Object {
          "id": "minValue",
          "preview": [Function],
          "render": [Function],
          "title": "Min value",
          "types": Array [
            "Currency",
            "Number",
            "Percentage",
            "Unit",
          ],
        },
        Object {
          "id": "maxValue",
          "preview": [Function],
          "render": [Function],
          "title": "Max value",
          "types": Array [
            "Currency",
            "Number",
            "Percentage",
            "Unit",
          ],
        },
      ]
    }
    onClose={[Function]}
    startingTabId={null}
    title="Number validation"
  />
</ContextProvider>
`;

exports[`AnswerValidation validation object array should render the Earliest Date validation 1`] = `
<Component
  displayName="Earliest date"
  readKey="earliestDate"
  testId="earliest-date-validation"
>
  <Component />
</Component>
`;

exports[`AnswerValidation validation object array should render the Latest Date validation 1`] = `
<Component
  displayName="Latest date"
  readKey="latestDate"
  testId="latest-date-validation"
>
  <Component />
</Component>
`;

exports[`AnswerValidation validation object array should render the Max Duration validation 1`] = `
<Component
  displayName="Max duration"
  readKey="maxDuration"
  testId="max-duration-validation"
>
  <Component />
</Component>
`;

exports[`AnswerValidation validation object array should render the Max value validation 1`] = `
<Component
  displayName="Max value"
  readKey="maxValue"
  testId="max-value-validation"
>
  <Component />
</Component>
`;

exports[`AnswerValidation validation object array should render the Min Duration validation 1`] = `
<Component
  displayName="Min duration"
  readKey="minDuration"
  testId="min-duration-validation"
>
  <Component />
</Component>
`;

exports[`AnswerValidation validation object array should render the Min value validation 1`] = `
<Component
  displayName="Min value"
  readKey="minValue"
  testId="min-value-validation"
>
  <Component />
</Component>
`;
